### PR TITLE
8334706: [JVMCI] APX registers incorrectly exposed on AMD64

### PR DESCRIPTION
**Pull request by dougxc**
This PR fixes a bug introduced by [JDK-8329032](https://bugs.openjdk.org/browse/JDK-8329032) which added the APX registers to `AMD64.java`.
It broke [this invariant](https://github.com/openjdk/jdk/blob/d2bebffb1fd26fae4526afd33a818ee776b7102e/src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/code/Architecture.java#L48-L52):
```
    /**
     * List of all available registers on this architecture. The index of each register in this list
     * is equal to its {@linkplain Register#number number}.
     */
    private final RegisterArray registers;
```
That invariant is relied upon by the Graal register allocator.
This PR now tests the invariant and fixes the definitions in `AMD64.java` that were violating it.

This fix was developed by @tkrodriguez.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334706](https://bugs.openjdk.org/browse/JDK-8334706): [JVMCI] APX registers incorrectly exposed on AMD64 (**Bug** - P2)

### Reviewers
 * [Yudi Zheng](https://openjdk.org/census#yzheng) (@mur47x111 - Committer) ⚠️ Review applies to [f8028db1](https://git.openjdk.org/jdk/pull/19824/files/f8028db188505224142221a96796e2f74eedac0e)
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19824/head:pull/19824` \
`$ git checkout pull/19824`

Update a local copy of the PR: \
`$ git checkout pull/19824` \
`$ git pull https://git.openjdk.org/jdk.git pull/19824/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19824`

View PR using the GUI difftool: \
`$ git pr show -t 19824`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19824.diff">https://git.openjdk.org/jdk/pull/19824.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19824#issuecomment-2182444826)

### CODE REVIEW

**Review by mur47x111 — APPROVED** [review]

**Comment by eme64** [comment]
@jatin-bhateja authored the changes in https://github.com/openjdk/jdk/pull/19042 that causes your issues, and @vnkozlov reviewed it. Maybe they can have a look at this?

File: src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/amd64/AMD64.java [hunk 1]
```diff
@@ -85,9 +85,7 @@ public class AMD64 extends Architecture {
 
     public static final Register[] cpuRegisters = {
         rax, rcx, rdx, rbx, rsp, rbp, rsi, rdi,
-        r8, r9, r10, r11, r12, r13, r14, r15,
-        r16, r17, r18, r19, r20, r21, r22, r23,
-        r24, r25, r26, r27, r28, r29, r30, r31
+        r8, r9, r10, r11, r12, r13, r14, r15
```
Comment (jatin-bhateja) [inline src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/amd64/AMD64.java:88]: Hi @dougxc , A comment over cpuRegisters definition relating to its usage will be helpful.

Comment (dougxc) [reply src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/amd64/AMD64.java:89]: I added a brief comment. I will leave it for [JDK-8334717](https://bugs.openjdk.org/browse/JDK-8334717) to expand upon this.

**Comment by dougxc** [comment]
Thanks for the review.

/integrate

**Review by tkrodriguez — APPROVED** [review]
Looks good.  We should clean AMD64.java up a bit in a follow on PR.  It should be more obvious how to make these kinds of changes without touching a bunch of arrays.

**Pull request closed without merging**